[WIP] Conda package

jonmmease opened the pull request:
Work towards adding a conda build of orca to the CI build.

Discussion: https://github.com/plotly/orca/issues/107
Standalone prototype: https://github.com/jonmmease/orca-feedstock

jonmmease commented:
@etpinard it doesn't look like the Travis build ran. Is there anything else I need to do?

etpinard commented:
> it doesn't look like the Travis build ran (I only see AppVeyor)

Ha. It might only run on branches based off the repo origin, and not on branches based off forks. 

I should get an invite giving you write rights to this repo. I think the easier way forward would be for you to pull down this repo, cherry-pick your commits unto a new branch and make a new PR. Sorry for the inconvenience.

etpinard commented:
Actually, pushing another commit to this branch should now do the trick: 

![image](https://user-images.githubusercontent.com/6675409/43587395-21aa3f14-9638-11e8-935a-6fe0896f5f20.png)

![image](https://user-images.githubusercontent.com/6675409/43587468-4a7c8aa0-9638-11e8-8e20-7799fbd4a3d3.png)

jonmmease commented:
Cool, I'll give that a try first

jonmmease commented:
Looks like they're all running now. Thanks!

jonmmease commented:
@etpinard the conda builds for Circle/Travis/AppVeyor are all working now. Each one successfully performs the conda build and lists the directory where the build artifact is output.

The `docker-build-and-push` job is failing on "Push Docker image to Quay", which is probably a good thing since I'm not intending up push a new docker image 🙂 

All that's left now is to figure out how we want to upload the packages to anaconda cloud.  I don't have a good sense of how this part should work.  Should the CI jobs do the upload, or should we try to get the built packages out of the CI systems and upload them manually?

Also, someone with some experience with these CI tools should probably take a look at what I did to get things working 🙂

etpinard commented on .circleci/config.yml:
```diff
@@ -21,6 +21,7 @@ workflows:
           requires:
             - test-node-v6
             - test-node-v8
+      - conda-build-and-upload
```
We probably only want to run this if `test-node-v6` and `test-node-v8` succeed. Adding 

```
  requires:
     - test-node-v6
     - test-node-v8
```

should do the trick

etpinard replied:
... and we should probably rename this step `conda-build` as we won't be uploading the builds on CI.

etpinard commented on .circleci/config.yml:
```diff
@@ -156,3 +157,29 @@ jobs:
+          keys:
+            - v3-npm-
+
+      - run: wget http://repo.continuum.io/miniconda/Miniconda3-latest-Linux-x86_64.sh -O miniconda.sh
```
We could probably cache this, but `conda-build-and-upload` runs in under 3 minutes, so I don't see the need :ok_hand:

etpinard commented on package.json:
```diff
@@ -1,6 +1,6 @@
 {
   "name": "orca",
-  "version": "1.0.0",
+  "version": "1.0.1-rc.1",
```
Did you base your branch off `master` or #103 ? If the former, we should stick to `1.0.0` here.

etpinard commented on recipe/meta.yaml:
```diff
@@ -0,0 +1,16 @@
+{% set data = load_setup_py_data() %}
```
We should add a README.md file in `recipe/` briefly explanation the conda-build process.

jonmmease commented:
@etpinard 
Well that took more tries that I thought it would 🙂 The artifact zip files look good for CircleCI and AppVeyor. I don't have access to the Travis artifacts (at least I don't have an AWS account), could you check that one?

I think I also addressed all of your feedback. Let me know if the Travis artifacts need any work, and if there are any other changes you'd like to see before merging this in.

Thanks!

etpinard commented:
> I don't have access to the Travis artifacts

Actually, looks like the artifacts didn't get pushed to S3 after your commits, perhaps due to another permissions restriction.

Oh well, I fetched your branch and created https://github.com/plotly/orca/compare/conda-build and all CI builds there (including the `docker-build-and-push` job on Circle) are :white_check_mark: 

Here's the link to Mac conda build:

https://s3.amazonaws.com/image-exporter-travis-artifacts/plotly/orca/433/433.1/conda-osx-64.zip

if you like to try it out.

etpinard commented on recipe/README.md:
```diff
@@ -0,0 +1,83 @@
+The resulting package will be named `plotly-orca-*.tar.bz2`, and the build
+command will display the full path of the output location. 
+
+How it works
```
Wow. I didn't expect this much detail. Thanks for your efforts!

jonmmease replied:
haha, sure. Once I got started a bunch of details came to mind that I thought would be useful background for future contributors.

etpinard commented on README.md:
```diff
@@ -14,6 +14,17 @@ Follow [@plotlygraphs](https://twitter.com/plotlygraphs) on Twitter for Orca ann
 
 ## Installation
 
+### Method 1: conda
```
Beautifully done. Thanks!

etpinard commented on CONTRIBUTING.md:
```diff
@@ -114,6 +114,19 @@ automate this process somehow.
   + Grab the Mac build automatically pushed to [Amazon S3](https://s3.console.aws.amazon.com/s3/buckets/image-exporter-travis-artifacts/plotly/orca/?region=us-east-1&tab=overview) from Travis.
     **N.B.** Select the latest build (largest number *note:* the folders are not necessarily sequential) -> release.zip
 - Run `npm publish`
+- Publish conda packages:
```
Very clear. :books:

jonmmease commented:
Replaced by https://github.com/plotly/orca/pull/113
